File: Zadanie4.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace UserPanelApp
{
    public class Permission
    {
        public string Name {get;set;}

        public Permission(string name)
        {
            Name = name;
        }

        public override string ToString()
        {
            return Name;
        }
    }

    public class Role
    {
        public string Name {get;set;}
        private List<Permission> permissions = new List<Permission>();

        public Role(string name)
        {
            Name = name;
        }

        public void AddPermission(Permission permission)
        {
            if (!permissions.Contains(permission))
            {
                permissions.Add(permission);
            }
        }

        public void RemovePermission(Permission permission)
        {
            permissions.Remove(permission);
        }

        public List<Permission> GetPermissions()
        {
            return permissions;
        }

        public override string ToString()
        {
            return Name;
        }
    }

    public class User
    {
        public string Username {get;set;}
        private List<Role> roles = new List<Role>();

        public User(string username)
        {
            Username = username;
        }

        public void AddRole(Role role)
        {
            if (!roles.Contains(role))
            {
                roles.Add(role);
            }
        }

        public void RemoveRole(Role role)
        {
            roles.Remove(role);
        }

        public List<Role> GetRoles()
        {
            return roles;
        }

        public bool HasPermission(string permissionName)
        {
            return roles
                .SelectMany(r => r.GetPermissions())
                .Any(p => p.Name == permissionName);
        }

        public void DisplayInfo()
        {
            Console.WriteLine($"Użytkownik: {Username}");

            foreach (var role in roles)
            {
                Console.WriteLine($"  Rola: {role.Name}");
                foreach (var perm in role.GetPermissions())
                {
                    Console.WriteLine($"    Uprawnienie: {perm.Name}");
                }
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Permission read = new Permission("READ");
            Permission write = new Permission("WRITE");
            Permission delete = new Permission("DELETE");

            Role admin = new Role("Admin");
            admin.AddPermission(read);
            admin.AddPermission(write);
            admin.AddPermission(delete);

            Role writer = new Role("Writer");
            writer.AddPermission(read);
            writer.AddPermission(write);

            Role viewer = new Role("Viewer");
            viewer.AddPermission(read);

            User user1 = new User("User1");
            User user2 = new User("User2");

            user1.AddRole(admin);
            user2.AddRole(viewer);

            user1.DisplayInfo();
            user2.DisplayInfo();

            Console.WriteLine();
            Console.WriteLine($"Czy User1 ma DELETE? {user1.HasPermission("DELETE")}");
            Console.WriteLine($"Czy User2 ma WRITE? {user2.HasPermission("WRITE")}");

            user1.RemoveRole(admin);
            Console.WriteLine("\nPo usunięciu roli Admin:");
            user1.DisplayInfo();
        }
    }
}

File: Zadanie4.py
class Permission:
    def __init__(self, name: str):
        self.name = name

    def __repr__(self):
        return f"Permission({self.name})"

    def __eq__(self, other):
        return isinstance(other, Permission) and self.name == other.name

    def __hash__(self):
        return hash(self.name)


class Role:
    def __init__(self, name: str):
        self.name = name
        self.permissions: Set[Permission] = set()

    def add_permission(self, permission: Permission):
        self.permissions.add(permission)

    def remove_permission(self, permission: Permission):
        self.permissions.discard(permission)

    def has_permission(self, permission: Permission) -> bool:
        return permission in self.permissions

    def __repr__(self):
        return f"Role({self.name})"


class User:
    def __init__(self, username: str):
        self.username = username
        self.roles: Set[Role] = set()
        self.direct_permissions: Set[Permission] = set()

    def add_role(self, role: Role):
        self.roles.add(role)

    def remove_role(self, role: Role):
        self.roles.discard(role)

    def add_permission(self, permission: Permission):
        self.direct_permissions.add(permission)

    def remove_permission(self, permission: Permission):
        self.direct_permissions.discard(permission)

    def has_permission(self, permission: Permission) -> bool:
        if permission in self.direct_permissions:
            return True

        for role in self.roles:
            if role.has_permission(permission):
                return True

        return False

    def __repr__(self):
        return f"User({self.username})"

if __name__ == "__main__":
    read = Permission("read")
    write = Permission("write")
    delete = Permission("delete")

    admin = Role("admin")
    writer = Role("writer")
    viewer = Role("viewer")

    admin.add_permission(read)
    admin.add_permission(write)
    admin.add_permission(delete)

    writer.add_permission(read)
    writer.add_permission(write)

    viewer.add_permission(read)

    user1 = User("User1")
    user2 = User("User2")

    user1.add_role(admin)
    user2.add_role(viewer)
    
    print(user2, "Czy może wpisać treść?", user2.has_permission(write))

    user2.add_permission(write)

    print(user1, "Czy może usuwać treści?", user1.has_permission(delete))
    print(user2, "Czy może wpisać treść?", user2.has_permission(write))
    print(user2, "Czy może usunąć treść?", user2.has_permission(delete))

    user1.remove_role(admin)
    print(user1, "Czy może usunąć treść po usunięciu roli?", user1.has_permission(delete))